### PR TITLE
[2.3]creating customer without password is directly confirmed 14492

### PR DESCRIPTION
**Pull request by novikor**
### Description (*)
Now, customer account is not confirmed when created without password;
Account IS confirmed if customer follows create/forgot password link
### Fixed Issues (if relevant)

1. magento/magento2#14492: Creating Customer without password is directly confirmed

### Manual testing scenarios (*)

**Admin Panel**

0. 'Require Emails Confirmation' is set to 'Yes'.
1. Admin creates account for the Customer in Admin Panel.
2. The Customer receives email with ask to change password for newly created account.
3. The Customer does not receive email for account confirmation.
4. The Customer successfully changes password for the account. The account becomes confirmed.
5. The Customer tries to login.
6. The Customer able to login for his/her account.

**Frontend**
_Case 1_

0. 'Require Emails Confirmation' is set to 'Yes'.
1. The Customer registers account with frontend registration.
2. The Customer receives confirmation email.
3. The Customer does not confirm account.
4. The Customer tries to login.
6. The Customer can not login and receive message 'This account is not confirmed. Click here to resend confirmation email.'.
7. The Customer tries to reset his/her password with'forgotpassword' functionality.
8. The Customer receives email for password reset procedure and use link from email.
9. The Customer changes his/her password. The account becomes confirmed.
10. The Customer tries to login.
11. The Customer able to login for his/her account.

_Case 2_

0. 'Require Emails Confirmation' is set to 'Yes'.
1. The Customer registers account with frontend registration.
3. The Customer receives email for account confirmation and use link from email.
4. The Customer confirms account.
5. The Customer tries to login.
6. The Customer able to login for his account.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by magento-engcom-team** [comment]
Hi @novikor. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

File: app/code/Magento/Customer/Controller/Account/CreatePassword.php [hunk 1]
```diff
@@ -34,20 +36,30 @@ class CreatePassword extends \Magento\Customer\Controller\AbstractAccount implem
     protected $resultPageFactory;
 
     /**
-     * @param Context $context
-     * @param Session $customerSession
-     * @param PageFactory $resultPageFactory
-     * @param AccountManagementInterface $accountManagement
+     * @var \Magento\Customer\Model\ForgotPassword|null
+     */
+    private $forgotPasswordModel;
```
Comment (swnsma) [inline app/code/Magento/Customer/Controller/Account/CreatePassword.php]: I think there can be better naming for class field.

File: app/code/Magento/Customer/Model/AccountManagement.php [hunk 2]
```diff
@@ -339,6 +338,11 @@ class AccountManagement implements AccountManagementInterface
      */
     private $addressRegistry;
 
+    /**
+     * @var \Magento\Customer\Model\ForgotPassword|null
+     */
+    private $forgotPasswordModel;
```
Comment (swnsma) [inline app/code/Magento/Customer/Model/AccountManagement.php]: I think there can be better naming for class field.

File: app/code/Magento/Customer/Model/ForgotPassword.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+use Magento\Framework\Exception\State\ExpiredException;
+use Magento\Framework\Phrase;
+
+/**
+ * Class ForgotPassword
+ *
+ * @package Magento\Customer\Model
+ * @SuppressWarnings(PHPMD.LongVariable)
+ */
+class ForgotPassword
```
Comment (swnsma) [inline app/code/Magento/Customer/Model/ForgotPassword.php]: Why it names 'Forgot Password' and:
1. Search customer by token
2. Confirm customer 

Can we make this SOLID?

**Review by swnsma — CHANGES_REQUESTED** [review]
Hi @novikor,
1. Please check my comments.
2. Please check failed tests.

File: app/code/Magento/Customer/Model/ForgotPassword.php (hunk 3 above)
Comment (novikor) [reply app/code/Magento/Customer/Model/ForgotPassword.php]: Done

**Comment by novikor** [comment]
> Hi @novikor,
> 
> 1. Please check my comments.
> 2. Please check failed tests.

Done. Please recheck.

File: app/code/Magento/Customer/Model/ForgotPasswordToken/ConfirmByToken.php [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+ * Copyright © Magento, Inc. All rights reserved.
+ * See COPYING.txt for license details.
+ */
+
+namespace Magento\Customer\Model\ForgotPasswordToken;
+
+use Magento\Customer\Api\CustomerRepositoryInterface;
+
+/**
+ * Class ConfirmByToken
```
Comment (swnsma) [inline app/code/Magento/Customer/Model/ForgotPasswordToken/ConfirmByToken.php]: Repeating class name in the comment is redundant. Clear responsibility description is sufficient.

File: app/code/Magento/Customer/Model/ForgotPasswordToken/GetByToken.php [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+
+use Magento\Customer\Api\CustomerRepositoryInterface;
+use Magento\Customer\Api\Data\CustomerInterface;
+use Magento\Framework\Api\SearchCriteriaBuilder;
+use Magento\Framework\Exception\NoSuchEntityException;
+use Magento\Framework\Exception\State\ExpiredException;
+use Magento\Framework\Phrase;
+
+/**
+ * Class GetByToken
```
Comment (swnsma) [inline app/code/Magento/Customer/Model/ForgotPasswordToken/GetByToken.php]: Repeating class name in the comment is redundant. Clear responsibility description is sufficient.

File: app/code/Magento/Customer/Model/ForgotPasswordToken/GetByToken.php [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+ *
+ * @package Magento\Customer\Model\ForgotPasswordToken
+ * @SuppressWarnings(PHPMD.LongVariable)
+ */
+class GetByToken
```
Comment (swnsma) [inline app/code/Magento/Customer/Model/ForgotPasswordToken/GetByToken.php]: Maybe, better naming will be GetCustomerByToken or Customer\GetByToken

**Review by swnsma — CHANGES_REQUESTED** [review]
Please proceed with minor changes.

File: app/code/Magento/Customer/Model/ForgotPasswordToken/ConfirmByToken.php [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+    /**
+     * Confirm customer account my rp_token
+     *
+     * @param string $resetPasswordToken
+     *
+     * @return void
+     * @throws \Magento\Framework\Exception\LocalizedException
+     */
+    public function execute(string $resetPasswordToken)
```
Comment (swnsma) [inline app/code/Magento/Customer/Model/ForgotPasswordToken/ConfirmByToken.php]: Please add return type.

File: app/code/Magento/Customer/Model/ForgotPasswordToken/ConfirmByToken.php (hunk 4 above)
Comment (novikor) [reply app/code/Magento/Customer/Model/ForgotPasswordToken/ConfirmByToken.php]: Cannot agree, but fixed.

File: app/code/Magento/Customer/Model/ForgotPasswordToken/GetByToken.php (hunk 5 above)
Comment (novikor) [reply app/code/Magento/Customer/Model/ForgotPasswordToken/GetByToken.php]: Cannot agree, but fixed.

File: app/code/Magento/Customer/Model/ForgotPasswordToken/GetByToken.php (hunk 6 above)
Comment (novikor) [reply app/code/Magento/Customer/Model/ForgotPasswordToken/GetByToken.php]: Done

File: app/code/Magento/Customer/Model/ForgotPasswordToken/ConfirmByToken.php (hunk 7 above)
Comment (novikor) [reply app/code/Magento/Customer/Model/ForgotPasswordToken/ConfirmByToken.php]: Done

**Comment by novikor** [comment]
@swnsma , please review

**Review by swnsma — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @swnsma, thank you for the review.
[ENGCOM-4581](https://jira.corp.magento.com/browse/ENGCOM-4581) has been created to process this Pull Request

**Comment by VasylShvorak** [comment]
✔️ QA passed

**Comment by shikhamis11** [comment]
@magento-engcom-team give me test instance

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11. Thank you for your request. I'm working on Magento instance for you

**Comment by magento-engcom-team** [comment]
Hi @shikhamis11, here is your [new Magento instance](https://pr-21394.instances.magento-community.engineering).
Admin access: [https://pr-21394.instances.magento-community.engineering/admin](https://pr-21394.instances.magento-community.engineering/admin)
Login: `admin` Password: `123123q`

**Comment by sivaschenko** [comment]
Hi @novikor can you please update your branch and resolve conflicts

**Comment by novikor** [comment]
Hi, @sivaschenko . Sorry for the delay.
The branch is up to date with 2.3-develop now.

**Comment by VasylShvorak** [comment]
:heavy_check_mark: QA passed

**Pull request merged**